Update Dockerfile for Python 3 and latest Balena and OctoPrint changes

RyanBalfanz opened the pull request:
These changes enable installed plugin persistence to the data directory, as well as the OctoPrint `config.yaml`.

In addition, it focuses on Python 3. The latest OctoPrint (v.14.0) [adds Python 3 support  as Python 2 is EOL](https://github.com/OctoPrint/OctoPrint/releases/tag/1.4.0).

## Requirements?

- [x] Python 3 ready
- [x] Ability to install plugins reliably (i.e. persistent, and resilient to system reboot and shutdown)
Note that restarting the application via a container restart is available but not the target in the UI because the is not equivalent to a reboot of the device. A "restart" in the UI is really a reboot for this reason. You can restart the application using the `balenactl restart` if you really want to start with a fresh container. This has implications with how the `config.yaml` file is currently managed during the first run. The config option to restart is defined for the convenience offered (e.g. easy "restart" via post-plugin install popup button).
- [x] Ability to create **and restore** from backups
- [ ] Ability to restart the octoprint server without restarting the container

## Nice to have:

### Various Balena application and device [Actions](https://www.balena.io/docs/learn/manage/actions/)

- [ ] Ability to toggle public device URL from within OctoPrint
- [ ] Ability to purge data directory from within OctoPrint
- [ ] [Update locking](https://www.balena.io/docs/learn/manage/actions/#update-locking), especially while printing is in progress, or during other OctoPrint operations such as installing plugins, restoring from backup, etc.

CC: @curcuz

Commit:
Update Dockerfile for Python 3 and latest Balena and OctoPrint changes

RyanBalfanz committed:
Replace scripts with new balenactl

RyanBalfanz committed:
!fixup remove dead code

RyanBalfanz committed:
Simplify start.sh script

RyanBalfanz committed:
Split long octoprint command to multiple lines

RyanBalfanz committed:
Use a subdirectory of /data for OctoPrint basedir

RyanBalfanz commented:
This information belongs here just as much, if not more?, than in the other PR:

https://github.com/balena-io-playground/balena-octoprint/pull/4#issuecomment-617439199

RyanBalfanz commented:
Still missing is the ability to restart the octoprint server *without* restarting the container.

RyanBalfanz committed:
Configure OctoPrint from octoprint config

RyanBalfanz committed:
!fixup typo

RyanBalfanz committed:
Do not install build-essential

RyanBalfanz committed:
!fixup typo

LongLiveCHIEF commented on Dockerfile.template:
```diff
@@ -1,46 +1,28 @@
-	&& cd /usr/src/app/octoprint && pip install "https://github.com/OctoPrint/OctoPrint-DisplayProgress/archive/master.zip" \
-	&& cd /usr/src/app/octoprint && pip install https://github.com/malnvenshorn/OctoPrint-CostEstimation/archive/master.zip
+# Install OctoPrint from source
+RUN git clone https://github.com/foosel/OctoPrint.git \
```
Needs to be `https://github.com/OctoPrint/OctoPrint`

RyanBalfanz replied:
Ok, that may be more correct nowadays. It seems that the repo was moved at some point. The foosel/OctoPrint does properly redirect, however, so this still works which I why I never changed it.

RyanBalfanz replied:
Ok, this is done.

LongLiveCHIEF replied:
It was only just moved 2 weeks ago, so it was easy to miss!

RyanBalfanz committed:
Clone OctoPrint from new OctoPrint/OctoPrint URL

Since https://github.com/foosel/OctoPrint.git was previously moved to https://github.com/OctoPrint/OctoPrint.git redirects still work here but this is more correct.

thebigredgeek commented:
@RyanBalfanz why don't we branch this into a different project?  Would be more than happy to help maintain it, too

RyanBalfanz commented:
Hi, I actually had recently been thinking to just keep my personal fork as
my main repo, rather than wait for merges here and then pull back to into
mine.

What did you have in mind?

On Thu, May 14, 2020 at 6:54 PM Andrew E. Rhyne <notifications@github.com>
wrote:

> @RyanBalfanz <https://github.com/RyanBalfanz> why don't we branch this
> into a different project? Would be more than happy to help maintain it, too
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/balena-io-playground/balena-octoprint/pull/5#issuecomment-628979487>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABARHX2UNHM7GHNVM67DSLRRSOEDANCNFSM4MM67FVQ>
> .
>
-- 
Sent from Gmail Mobile

RyanBalfanz commented:
@thebigredgeek: Just to follow up from my previous message, wanted to pass along these other projects as well.

1. https://github.com/AmedeeBulle/octoprint-containers
2. https://github.com/OctoPrint/docker

As for (1), it seems pretty good. The only downside that I think exists is that multi-container support is a paid feature with Balena.io. I beleive it should be possible to create a complete solution (e.g. OctoPrint, webcam support, etc.) using Supervisor, for example, in single container if need be.

As for (2), @LongLiveCHIEF can speak more to it than I can.

RyanBalfanz commented:
@thebigredgeek, @LongLiveCHIEF 

Following up from my last comment–I just created an example showing how to use Supervisor with Balena. In this example the web port 80 is dedicated to the Supervisor internet interface, but can be swapped for OctoPrint instead. Supervisor only needs to be used to run (a) OctoPrint, (b, c…) auxiliary applications such as those to support the webcam.

The code is at https://github.com/RyanBalfanz/balena-supervisor.

brainstorm commented:
I would actually prefer to see this merged on `balena-io` instead, seems to be more findable instead of hunting for different individual repos?

curcuz commented:
@RyanBalfanz sorry for the long time it took to notice your PR, but it's still marked as WIP :P

Is this ready for review? 

Also, I'm looking within the Balena team for a maintainer to promote this project from experiment playground (as it is now under the `balena-io-playground` org) to maintained project so hopefully there will be better response to community in the near future.

curcuz commented:
@RyanBalfanz @brainstorm @thebigredgeek I was pointed by colleagues that there is already a well-maintained community project here https://github.com/MatthewCroughan/octobalena that is actually a fork of this one. 

I am considering deprecating this project and directing people towards octobalena.

Fun fact: Matthew, the maintainer, was also our guest in  our [IoT Happy Hour EP9](youtube.com/watch?v=0Wew51n0KV0)

chrisys commented:
I've updated this project and kept it as a barebones version, but octobalena linked above is still a great option for a more fully featured version. There are definitely still some nice ideas in this PR we could implement in a basic version though.

Just to clarify though, you can still run multicontainer applications on the free tier of balenaCloud - that's exactly what octobalena is doing :)
